### ISSUE TITLE
Planning - Part 2: What does the service provide?

### ISSUE THREAD

**Issue by jamen** [soundcove/info#2]
Can we discus what this service actually provides in-depth? We need to know the fine details of what our service does before we can design a API around it.

To start us off.  I know that our service provides registration - quite simply you can register an account with our service.  An account on our service would contain a login name, display name, profile picture, a short bio, and possibly links to other sites you're registered on (anything else?).  Along-side typical account info, you can also have playlists, which are collections of songs (from YouTube, SoundCloud, possibly others?).

If you're unsure about a feature, post it anyways and we can discuss it.

**Comment by jamen** [comment]
### Groups

Groups can be private or public depending on what the creator of the group wants.

#### Terminology 
- **Hosts** are the creator of the given group.
- **Mods or Moderators** moderate the group and users, they are given this role by the host.
- **Listeners** refer to people in the group that is not actively participating in the music sharing process.
- **DJs** refer to the people who are actively participating in the music sharing process. 
- **The DJ** refers to the currently active DJ in the cycle.
- **Chatters** refers to the users that are actively participating in the chat of the group.
- **The wait-list** refers to the list you're placed on when you decide to "DJ" (participate in the music sharing process).

The host can create and revoke users' moderator roles.  Moderators have the same power as the host besides a select few actions, like creating / revoking moderator roles and deleting the group.

Rooms contain chats in which you can communicate with other users that are in the group.  The host and mods of the groups are specially marked in the chat (perhaps with a badge or separating color).  The host and mods can also mute users in the chat for a certain duration or forever.  The chat would have support for emoticons (perhaps much like GitHub or Slack's system).

When a user decides to DJ, they are placed at the very end of the wait-list.  The system cycles through each member in the wait-list playing one song from their selected playlist at a time; however, the host and moderators can set restrictions for DJs, such as if there is a certain percentage or amount of skip votes then the cycle will continue to the next user.  Hosts and moderators can also skip the to next user directly (such as if the song is inappropriate).  Hosts and moderators can also set song duration limits.  If the song is too long or too short you can be skipped.  You can also be banned from DJing (you'll be kicked from the wait-list).

**Comment by programmatical** [comment]
For the Youtube and Soundcloud extensions, I think it would be great to have an initial authentication throughout the app and just store that auth instance, so the user can access his/her playlists and add them directly if they want.

**Comment by programmatical** [comment]
1. Some questions to consider:

2. Do we want to use a simply lobby system, lobbies that are not persistent?

3. Do we want to host persistent rooms ourselves?

4. Should users be able to create persistent rooms by default or should payment and/or verification be involved?

5. If we consider #4 should we allow them to host rooms themselves independently?

6. Can users host multiple rooms?

Helpful Terms:

- persistent: Should rooms last for ever or should we develop a pruning system that removes them automatically after a specific criteria has been met?

**Comment by jamen** [comment]
I do agree.  To elaborate on the lobby expiration system, I'm thinking it will probably be on a time basis.  We had discussed the minimum and decided around 3 months of inactivity declares a lobby inactive, however inactivity checks are weekly.  

Now, more on "hubs", the plan was to have these created through a graduation system.  If your lobby has a high amount of listeners and DJs consistently, then over time you'll be given the option to graduate your lobby into a hub.  Hubs cannot be pruned from the lobby expiration system...  We haven't decided the minimum requirements for a lobby to become a hub, any ideas?

**Comment by jamen** [comment]
I think with what we have planned so far, we have a pretty good foundation of what needs to go into the API.  Closing this issue and creating a new issue for the API structure.

If we need more foundation to it later on, I'll open another issue and we can discus.
